### ISSUE TITLE
[java] UnusedPrivateField false-positive

### ISSUE THREAD

**Issue by cynthux** [pmd/pmd#1544]
**Affects PMD Version:** 6.11.0

**Rule:** UnusedPrivateField

**Description:**

False positive when instance variable is initialized in the constructor and no getter or setter for the field exists but it is used in GsonWrapper.toJson().

**Code Sample demonstrating the issue:**

```java
public class Point {
    private double x;
    private double y;

    public Point(double x, double y) {
        this.x = x;
        this.y = y;
    }
}

// In another class
Point point = new Point(1.5, 2.5);
String body = GsonWrapper.toJson(point);
```

**Running PMD through:** *Maven*

**Comment by jsotuyod** [comment]
@cynthux this is not within the scope of PMD for a number of reasons.

Firstly, because PMD analyses per-file, which means when we look into `Point` the fields are effectively unused.

Second, because even if we did multi-file analysis, we can't tell what `GsonWrapper.toJson` does (and even if we did, we would probably find a `gson.toJson` call within which leads to the same point), the fact that it's using **reflection** to access all fields is unknown from outside. That's semantic, and not a syntactic; and outside what we can feasibly achieve with static analysis. At the very best we would have to do some kind of whitelisting, which is subpar without proper customization...

Which leads us to the fact, we **can** already flag and whitelist classes we semantically know we want to use this way; and this can be done a number of ways:

1. You can have this rule ignore certain classes or packages you may want. For instance, to suppress this rule from all classes within any package ending in `.dto` you can:
```xml
<rule ref="category/java/bestpractices.xml/UnusedPrivateField">
  <properties>
    <property name="violationSuppressXPath" value="//PackageDeclaration/Name[ends-with(@Image, '.dto')]"/>
  </properties>
</rule>
```
2. You can create any custom annotation (ie: `@com.mercadolibre.Dto`), annotate the classes you need with it, and then have classes annotated with that class ignored:
```xml
<rule ref="category/java/bestpractices.xml/UnusedPrivateField">
  <properties>
    <property name="ignoredAnnotations" value="com.mercadolibre.Dto"/>
  </properties>
</rule>
```

More info on suppression for any rule can be found [here](https://pmd.github.io/pmd-6.10.0/pmd_userdocs_suppressing_warnings.html#the-property-violationsuppressxpath), while ignored annotations for this particular
rule can be found [here](https://pmd.github.io/pmd-6.10.0/pmd_rules_java_bestpractices.html#unusedprivatefield)

If you have any doubts, feel free to reach out to me on Hangouts (I work @ meli too!)
